### ISSUE TITLE
ExDoc cannot deal with literal map typespecs :optional and :required fields 

### ISSUE THREAD

**Issue by eksperimental** [elixir-lang/ex_doc#1219]
Can't deal with

```elixir
      | %{required(key_type) => value_type}   # map with required pairs of key_type and value_type
      | %{optional(key_type) => value_type}   # map with optional pairs of key_type and value_type
```

It assumes `required` and òptional` are types

**Comment by josevalim** [comment]
Can you please explain what you mean by "can't deal"? Do you get an error? If so, which one? Is it a warning? A failure to render? Thank you.

**Comment by eksperimental** [comment]
It warns about `t:optional/1` and `t:required\1` not being defined.
this commit in this pending PR solves it.
https://github.com/elixir-lang/ex_doc/pull/1233/commits/16939dfdf6982f7ec50dd776f0d92b5e49551674

**Comment by josevalim** [comment]
Thanks, I see! Maybe we should send a separate PR for it with tests. In particular, I would probably handle them in a separate clause. :)

**Comment by eksperimental** [comment]
Sorry, it does not work in master. It does in that PR that we fine grain the warnings.

**Comment by wojtekmach** [comment]
Fixed in #1233.
